witchcraft-log-util/src/filter.rs

```diff
@@ -0,0 +1,154 @@
+// Copyright 2025 Palantir Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//! A prefix-based target filter.
+
+use sequence_trie::SequenceTrie;
+use witchcraft_log::{LevelFilter, Metadata};
+
+/// A prefix-based target filter.
+///
+/// The filter is configured with a top-level [`LevelFilter`] and additional per-target filters. Targets are interpreted
+/// as a hierarchy by splitting on `::`. For example a target `foo::bar` will have a filter for the `foo` target
+/// applied to it if there is not also a filter for `foo::bar` itself.
+pub struct Filter {
+    trie: SequenceTrie<String, LevelFilter>,
+}
+
+impl Filter {
+    /// Returns a new builder.
+    #[inline]
+    pub fn builder() -> Builder {
+        Builder {
+            filter: Filter {
+                trie: SequenceTrie::new(),
+            },
+            root: LevelFilter::Error,
+        }
+    }
+
+    /// Determines if the provided log metadata matches the filter.
+    pub fn enabled(&self, metadata: &Metadata<'_>) -> bool {
+        metadata.level()
+            <= *self
+                .trie
+                .get_ancestor(metadata.target().split("::"))
+                .unwrap()
+    }
```

Comment:
This seems fairly expensive on a per-log basis. In the java implementation we have a split between the input configuration, and the state used by individual loggers (which map to origins) where we pre-compute level and filter information. In this case, the origin is always a provided value, so we can't get clever with effectively-free level/filter checks. Possibly worth considering in the future, but I admit I haven't fully wrapped my head around this library yet.

Reply:
I think it probably isn't a huge problem in practice though - if you're logging enough that per-invocation routing is a problem, you're probably already overwhelming your logging infrastructure. Extremely verbose trace-style logs get short circuited via the https://docs.rs/witchcraft-log/latest/witchcraft_log/fn.max_level.html: https://docs.rs/witchcraft-log/latest/src/witchcraft_log/macros.rs.html#20

Reply:
It's definitely a micro-optimization :-)